lmorpho/morphio.cpp

```diff
@@ -0,0 +1,173 @@
+#include <fstream>
+#include <iterator>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <swcio.hpp>
+
+#include "morphio.hpp"
+
+using nest::mc::io::swc_record;
+
+std::vector<swc_record> as_swc(const morphology& morph);
+
+// printf wrappers.
+
+template <typename... Args>
+static std::string strprintf(const char* fmt, Args&&... args) {
+    thread_local static std::vector<char> buffer(1024);
+
+    for (;;) {
+        int n = std::snprintf(buffer.data(), buffer.size(), fmt, std::forward<Args>(args)...);
+        if (n<0) return ""; // error
+        if ((unsigned)n<buffer.size()) return std::string(buffer.data());
+        buffer.resize(2*n);
+    }
+}
+
+template <typename... Args>
+static std::string strprintf(std::string fmt, Args&&... args) {
+    return strprintf(fmt.c_str(), std::forward<Args>(args)...);
+}
+
+// Multi-file manager implementation.
+multi_file::multi_file(const std::string& pattern, int digits) {
+    auto npos = std::string::npos;
+
+    file_.exceptions(std::ofstream::failbit);
+    concat_ = (pattern.find("%")==npos);
+    use_stdout_ = pattern.empty() || pattern=="-";
+
+    if (!concat_) {
+        std::string nfmt = digits? "%0"+std::to_string(digits)+"d": "%d";
+        std::string::size_type i = 0;
+        for (;;) {
+            auto p = pattern.find("%", i);
+
+            if (p==npos) {
+                fmt_ += pattern.substr(i);
+                break;
+            }
+            else {
+                fmt_ += pattern.substr(i, p-i);
+                fmt_ += i==0? nfmt: "%%";
+                i = p+1;
+            }
+        }
+    }
+    else {
+        filename_ = pattern;
+    }
+}
+
+void multi_file::open(unsigned n) {
+    if (use_stdout_ || (file_.is_open() && (concat_ || n==current_n_))) {
+        return;
+    }
+
+    if (file_.is_open()) file_.close();
+
+    std::string fname = concat_? filename_: strprintf(fmt_, n);
+    file_.open(fname);
+
+    current_n_ = n;
+}
+
+// SWC transform
+
+using nest::mc::io::swc_record;
+
+std::vector<swc_record> as_swc(const morphology& morph) {
+    using kind = swc_record::kind;
+    std::map<int, int> parent_end_id;
+    std::vector<swc_record> swc;
+
+    // soma
+    const auto &p = morph.soma;
+    int id = 0;
+    parent_end_id[0] = 0;
+    swc.emplace_back(kind::soma, id, p.x, p.y, p.z, p.r, -1);
+
+    // dendrites:
+    for (auto& sec: morph.sections) {
+        int parent = parent_end_id[sec.parent_id];
+
+        const auto& points = sec.points;
+        auto n = points.size();
+        if (n<2) {
+            throw std::runtime_error(strprintf("surprisingly short cable: id=%d, size=%ul", sec.id, n));
+        }
+
+        // include first point only for dendrites segments attached to soma.
+        if (sec.parent_id==0) {
+            const auto& p = points[0];
+            swc.emplace_back(kind::fork_point /*dendrite*/, ++id, p.x, p.y, p.z, p.r, parent);
```

Comment:
Its completely valid and it saves a line. But I would put the ++id on its own line, it feels like a side effect written like this, and they get forgotten